nixos/modules/system/boot/luksroot.nix

```diff
@@ -1081,6 +1095,44 @@ in
     boot.initrd.preLVMCommands = mkIf (!config.boot.initrd.systemd.enable) (commonFunctions + preCommands + concatStrings (mapAttrsToList openCommand preLVM) + postCommands);
     boot.initrd.postDeviceCommands = mkIf (!config.boot.initrd.systemd.enable) (commonFunctions + preCommands + concatStrings (mapAttrsToList openCommand postLVM) + postCommands);
 
+    boot.initrd.systemd.services = let devicesWithClevis = filterAttrs (device: _: (hasAttr device clevis.devices)) luks.devices; in
+      mkIf (clevis.enable && systemd.enable) ((mapAttrs'
+        (name: _: nameValuePair "cryptsetup-pre-clevis-${name}" {
+          requiredBy = [ "systemd-cryptsetup@${name}.service" "cryptsetup.target" "cryptsetup-pre.target" ];
+          before = [ "cryptsetup.target" "cryptsetup-pre.target" ];
+          wants = [ "systemd-udev-settle.service" "network-online.target" ];
+          after = [
+            "systemd-udev-settle.service"
+            "systemd-modules-load.service"
+          ];
+          unitConfig.DefaultDependencies = "no";
+          serviceConfig = {
+            Type = "oneshot";
+            RemainAfterExit = true;
+          };
+          script = ''
+            mkdir -p /clevis-${name}
+            mount -t ramfs none /clevis-${name}
+            clevis decrypt < /etc/clevis/${name}.jwe > /clevis-${name}/decrypted
+          '';
+        })
+        devicesWithClevis) // (mapAttrs'
+        (name: _: nameValuePair "cryptsetup-post-clevis-${name}" {
+          requires = [ "cryptsetup.target" ];
+          after = [
+            "cryptsetup.target"
+            "systemd-udev-settle.service"
+            "systemd-modules-load.service"
+            "systemd-ask-password-console.service"
+          ];
+          unitConfig.DefaultDependencies = "no";
+          serviceConfig.Type = "oneshot";
+          script = ''
+            umount /clevis-${name}
+          '';
+        })
+        devicesWithClevis));
```

Comment:
Rather than making this two different services, I'd just make this one service with an `ExecStop`, `conflicts = ["initrd-switch-root.target" "shutdown.target"];`, and `RemainAfterExit = true;`. Ironically, I *was* going to ask why you have `RemainAfterExit` in the service since it isn't doing anything as is, but this change would require it.